var/spack/repos/builtin/packages/py-pytorch/package.py

```diff
@@ -0,0 +1,93 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+
+class PyPytorch(PythonPackage):
+    """Tensors and Dynamic neural networks in Python
+    with strong GPU acceleration."""
+
+    homepage = "http://pytorch.org/"
+    url      = "https://github.com/pytorch/pytorch/archive/v0.3.1.tar.gz"
+
+    version('0.3.1', git='https://github.com/pytorch/pytorch.git',
+            tag='v0.3.1', submodules=True)
+    version('0.4.0', git='https://github.com/pytorch/pytorch.git',
+            tag='v0.4.0', submodules=True)
+
+    variant('cuda', default='False', description='Add GPU support')
+    variant('cudnn', default='False', description='Add cuDNN support')
+    variant('nccl', default='False', description='Add NCCL support')
+    variant('mkldnn', default='False', description='Add Intel MKL DNN support')
+    variant('magma', default='False', description='Add MAGMA support')
+
+    conflicts('+cudnn', when='~cuda')
+    conflicts('+nccl', when='~cuda')
+    conflicts('+magma', when='~cuda')
+    conflicts('+mkldnn', when='@:0.3.2')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-cffi', type='build')
+    depends_on('py-numpy', type=('run', 'build'))
+    depends_on('blas', type=('run', 'build'))
+    depends_on('lapack', type=('run', 'build'))
+    depends_on('py-pyyaml', type=('run', 'build'))
+    depends_on('py-typing', when='@0.3.2:', type=('run', 'build'))
+    depends_on('intel-mkl', when='+mkl', type=('build', 'run'))
+    depends_on('cuda', when='+cuda', type=('build', 'run'))
+    depends_on('cudnn', when='+cuda+cudnn', type=('build', 'run'))
+    depends_on('nccl', when='+cuda+nccl', type=('build', 'run'))
+    depends_on('magma+shared', when='+cuda+magma', type=('build', 'run'))
```

Comment:
Are you sure these should be `type=('build', 'run')`? For the Python packages it makes sense, but most of the rest of the dependencies are libraries. You should search the installation directory for `*.so` files and run `ldd -r` on them to see what they are linking to (or `*.dylib` and `otool -L` if on macOS).

Reply:
I left cuda since it appears in the ldflags needed by the JIT part.
They are all linked in a shared library so as you suggested it should be fine to remove the `'run'` dependencies

Reply:
> The other libraries are all linked in a shared library

So then shouldn't they be `type='link'` or the default `type=('build', 'link')`?

Reply:
Ping. I think these should be link-type dependencies.